tempesta_fw/http.c

```diff
@@ -382,6 +370,396 @@ tfw_http_send_504(TfwHttpReq *req)
+
+	/*
+	 * Process the server connection's queue of pending requests.
+	 * The queue is locked against concurrent updates: inserts of
+	 * outgoing requests, or closing of the server connection. Do
+	 * it as fast as possible by moving failed requests to other
+	 * queues that can be processed without the lock.
+	 */
+	req = srv_conn->msg_sent
+	    ? (TfwHttpReq *)list_next_entry(srv_conn->msg_sent, fwd_list)
+	    : (TfwHttpReq *)list_first_entry(fwd_queue, TfwMsg, fwd_list);
+
+	list_for_each_entry_safe_from(req, tmp, fwd_queue, msg.fwd_list) {
+		unsigned long jtimeout = jiffies - req->jtstamp;
+		if (time_after(jtimeout, srv->qjtimeout)) {
+			TFW_DBG2("%s: Eviction: req=[%p] overdue=[%dms]\n",
+				 __func__, req,
+				jiffies_to_msecs(jtimeout - srv->qjtimeout));
+			tfw_http_req_move2equeue(srv_conn, req, equeue, 504);
+			continue;
+		}
+		/*
+		 * If unable to send to the server connection due to
+		 * an error, then move the request to @err_queue for
+		 * sending a 500 error response later. That is safe
+		 * as the response will be sent in proper seq order.
+		 */
+		if (tfw_connection_send(srv_conn, (TfwMsg *)req)) {
+			TFW_DBG2("%s: Forwarding error: conn=[%p] req=[%p]\n",
+				 __func__, srv_conn, req);
+			tfw_http_req_move2equeue(srv_conn, req, equeue, 500);
+			continue;
+		}
+		srv_conn->msg_sent = (TfwMsg *)req;
+		/* Stop sending if the request is non-idempotent. */
+		if (tfw_http_req_is_nonidempotent(req)) {
+			TFW_DBG2("%s: Break on non-idempotent: req=[%p]\n",
+				 __func__, req);
+			break;
+		}
+		/* See if a non-idempotent request has become idempotent. */
+		tfw_http_req_nonidemp_delist(srv_conn, req);
+	}
+}
+
+/*
+ * Forward stalled requests in server connection @srv_conn.
+ *
+ * This function expects that the queue in the server connection
+ * is locked. The queue in unlocked inside the function which is
+ * very non-traditional. Please use with caution.
+ */
+static void
+tfw_http_req_fwd_stalled(TfwConnection *srv_conn)
+{
+	LIST_HEAD(equeue);
+
+	TFW_DBG2("%s: conn=[%p]\n", __func__, srv_conn);
+	WARN_ON(!spin_is_locked(&srv_conn->msg_qlock));
+	BUG_ON(list_empty(&srv_conn->msg_queue));
+
+	__tfw_http_req_fwd_stalled(srv_conn, &equeue);
+	spin_unlock(&srv_conn->msg_qlock);
+
+	if (!list_empty(&equeue))
+		tfw_http_req_zap_error(&equeue);
+}
+
+/*
+ * Forward the request @req to server connection @srv_conn.
+ *
+ * The request is added to the server connection (forwarding) queue.
+ * If forwarding is on hold at this moment, then the request will be
+ * forwarded later. Otherwise, if the queue is drained, then forward
+ * the request to the server immediately. If the queue is not drained,
+ * then forward all stalled requests to the server.
+ *
+ * Forwarding to a server is considered to be on hold after
+ * a non-idempotent request is forwarded to the server. The hold
+ * is removed when the holding non-idempotent request is followed
+ * by another request from the same client, which enables pipelining.
```

Comment:
Is that suits RFC 7230 6.3.2?
```
A user agent SHOULD NOT pipeline requests after a non-idempotent method, 
until the final response status code for that method has been received, unless 
the user agent has a means to detect and recover from partial failure conditions 
involving the pipelined sequence.
```
As i got it, we cannot remove hold untill request was answered